### ISSUE TITLE
oc project and oc login fail to allow switching to another server if the current server is down

### ISSUE THREAD

**Issue by smarterclayton** [openshift/origin#12343]
They fail to connect and therefore break.  This is a regression of `oc project` at least.

**Comment by juanvallejo** [comment]
@smarterclayton I cannot seem to replicate this issue in `oc version`:
```
oc v1.3.0-rc1+1cd60fb-1969-dirty
kubernetes v1.4.0+776c994
features: Basic-Auth

Server https://10.13.137.149:8443
openshift v1.3.0-rc1+1cd60fb-1969-dirty
kubernetes v1.4.0+776c994
```

I am able to use `oc login` when the current server is down to switch to another server:
```
$ oc cluster up
$ oc login --server=<localhost> -u system:admin
Logged into https://10.13.137.149:8443...
$ oc cluster down
$ oc whoami
The connection to the server 10.13.137.149:8443 was refused - did you specify the right host or port?

$ oc login --server=<new_server> --token=<my_token>
Logged into...
```

**Comment by smarterclayton** [comment]
this was when already logged in to both servers.  Login to both, take the one you are talking to down. Then try to login or switch projects.

**Comment by juanvallejo** [comment]
@smarterclayton I am still not able to reproduce the issue. Below are steps I am taking:
```
# log in to server1
$ oc login --server=<server1> --token=<token1>
Logged into "<server1>" as "sample-user" using the token provided.

You have access to the following projects and can switch between them with 'oc project <projectname>':

  * default
    sample-app

Using project "default".

# login to server 2
$ oc login --server=<server2> --token=<token2>
Logged into "<server2>" as "sample-user" using the token provided.

You have one project on this server: "sample-app"

Using project "sample-app".

# bring down server2
sudo kill <server2_pid>

# make sure <server2> is down
$ oc whoami
The connection to the server <server2> was refused - did you specify the right host or port?

# attempt to switch to a project in <server1>
$ oc project sample-app --context=<server1_context_name>
Now using project "sample-app" on server "<server1>".

# check current server address
$ oc whoami --show-server
<server1>

# check context
$ oc config current-context
<server1_context_name>
```

**oc login**
```
# log in to server1
$ oc login --server=<server1> --token=<token1>
Logged into "<server1>" as "sample-user" using the token provided.

You have access to the following projects and can switch between them with 'oc project <projectname>':

  * default
    sample-app

Using project "default".

# login to server 2
$ oc login --server=<server2> --token=<token2>
Logged into "<server2>" as "sample-user" using the token provided.

You have one project on this server: "sample-app"

Using project "sample-app".

# bring down server2
sudo kill <server2_pid>

# make sure <server2> is down
$ oc whoami
The connection to the server <server2> was refused - did you specify the right host or port?

# attempt to log in to <server1>
$ oc login --server=<server1> --token=<token1>
Logged into "<server1>" as "sample-user" using the token provided.

You have access to the following projects and can switch between them with 'oc project <projectname>':

    default
  * sample-app

Using project "sample-app".
```

**oc version**
```
oc v1.3.0-rc1+9877afb-2132-dirty
kubernetes v1.4.0+776c994
features: Basic-Auth

Server https://api.preview.openshift.com:443
openshift v3.3.1.3
kubernetes v1.3.0+52492b4
```

**Comment by smarterclayton** [comment]
It's likely that you will see different results when

1. the server has a DNS name that points to an invalid IP address
2. the server has a DNS name that is not an actual DNS name anymore
3. the server has an IP address and there is a firewall that blocks the IP
4. the server has an IP address and there is no firewall

You can pretty easily fake that by changing the server entry in your config
for the various scenarios.

**Comment by juanvallejo** [comment]
@fabianofranz @smarterclayton 
Tried the following steps but still unable to replicate this issue (using Fedora 23), apologies if I am still missing anything:

> 1. the server has a DNS name that points to an invalid IP address

```
# first, attempt to log in to <server1>
$ oc login --server=<server1> --token=<token>
Logged into "<server1>:443" as "juanvallejo" using the token provided.

You have one project on this server: "sample-app-2"

Using project "sample-app-2".
```
```
# then, add a dns entry "foobar" in /etc/hosts that resolves to "10.13.137.149", which is <server2>'s ip address
$ cat /etc/hosts
127.0.0.1               localhost.localdomain localhost
::1             localhost6.localdomain6 localhost6
10.13.137.149           foobar
```
```
# then, login to <server2> using <foobar> as the hostname
$ oc login https://foobar:8443 -u <testuser> -p <password> --insecure-skip-tls-verify
Login successful.

You have one project on this server: "sample-app-2"

Using project "sample-app-2".

# verify address of current context
$ oc whoami --show-server
https://foobar:8443
```
```
# verify that <server2> is the current context
$ oc config current-context
sample-app-2/foobar:8443/<testuser>

$ oc whoami --show-server
https://foobar:8443

# invalidate target ip address for the dns name "foobar" (previously pointed to "10.13.137.149", now points to "10.1.1.1"
$ cat /etc/hosts
127.0.0.1               localhost.localdomain localhost
::1             localhost6.localdomain6 localhost6
10.1.1.1                foobar

# verify that "foobar" is now unresolvable
$ ping foobar
PING foobar (10.1.1.1) 56(84) bytes of data.
^C
--- foobar ping statistics ---
4 packets transmitted, 0 received, 100% packet loss, time 3000ms
```
```
# with foobar pointing to an invalid IP address, attempt to login to <server1>
$ oc login --server=<server1> --token=<token>
Logged into "<server1>:443" as "juanvallejo" using the token provided.

You have one project on this server: "sample-app-2"

Using project "sample-app-2".
```

> the server has a DNS name that is not an actual DNS name anymore

By repeating the steps above, but deleting the "foobar" entry from `/etc/hosts` rather than invalidating the IP address it points to, I am still able to log in to `<server1>`

> the server has an IP address and there is a firewall that blocks the IP
```
# log into <server1> and <server2> using `oc login --server ...`
...

# verify that current context is <server2>
$ oc whoami --show-server
https://10.13.137.149:8443

# add an iptables rule to block <server2>'s ip address
$ sudo iptables -I INPUT 1 -s 10.13.137.149 -j DROP

# verify that <server2> is unreachable
$ 10.13.137.149
PING 10.13.137.149 (10.13.137.149) 56(84) bytes of data.
^C
--- 10.13.137.149 ping statistics ---
3 packets transmitted, 0 received, 100% packet loss, time 1999ms

# attempt to log in to <server1>
$ oc login --server=<server1> --token=<token>
Logged into "<server1>:443" as "juanvallejo" using the token provided.

You have one project on this server: "sample-app-2"

Using project "sample-app-2".
```

Please let me know if I missed anything, @fabianofranz could this maybe be replicated only on mac or windows?

**Comment by juanvallejo** [comment]
@smarterclayton I was able to replicate this issue by having my current context set to `<server_A>` (a server which only returns `500` errors), and then attempting to switch to a fully-working `<server_B>`:

```
# verify that current context is the "broken" server
$ oc config current-context
sample-app-2/<server_A>/system:admin

# attempt to switch contexts to (working server) <server_B> 
$ oc project <server_B_context>

error: an error on the server ("I only return error 500 responses.") has prevented the request from succeeding
See 'oc project -h' for help and examples.

# verify that context did not switch due to previous error
$ oc config current-context
<server_A_context>
```

I will open a PR to address this.
